Sprint 12

AntonDemidenk0 opened the pull request.

AntonDemidenk0 committed:
Создан сетевой слой для лента фотографий

AntonDemidenk0 committed:
создан сервис ImagesListService, который осуществляет постраничную загрузку данных, необходимых для ленты фотографий.

AntonDemidenk0 committed:
✅ в таблице используются не моковые картинки, а изображения из сети

AntonDemidenk0 committed:
- добавлен функционал лайков
- добавлена возможность просмотра полноразмерной картинки в ленте
- улучшен процесс работы кнопки логаута

AntonDemidenk0 committed:
Выполнено задание по 12 спринту

AntonDemidenk0 committed:
пропал progressView при загрузке вебстраницы логина, пофиксил.

Helvetii951 commented on AppDelegate.swift:
```diff
@@ -8,6 +9,9 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
         // Override point for customization after application launch.
+        ProgressHUD.animationType = .activityIndicator
+        ProgressHUD.colorHUD = .black
+        ProgressHUD.colorAnimation = .lightGray
```
Такие настройки лучше выносить в отдельные функции

Helvetii951 commented on Constants.swift:
```diff
@@ -84,4 +84,47 @@ enum Constants {
         request.httpMethod = "GET"
         return request
     }
+    static func makePhotosRequest(page: Int) -> URLRequest? {
+        guard let baseURL = defaultBaseURL else {
+            print("Invalid base URL")
+            return nil
+        }
+        let urlString = "/photos?page=\(page)"
+        guard let url = URL(string: urlString, relativeTo: baseURL) else {
+            print("Invalid URL string")
+            return nil
+        }
+        
+        var request = URLRequest(url: url)
+        if let token = Constants.tokenStorage.token {
+            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
+        } else {
+            print("Token not found")
+            return nil
+        }
+        
+        request.httpMethod = "GET"
+        return request
+    }
```
Формирование `URLRequest` стоит перенести в сервис, который использует этот запрос. Пусть в файле с константами будут только константы :)

Helvetii951 commented on ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,129 @@
+    func fetchPhotosNextPage() {
+        guard !isFetching, task == nil else { return }
+        isFetching = true
+        
+        nextPage += 1
+        
```
API Unsplash позволяет загружать фото с нулевой страницы(самый первый запрос будет с первой). Не суперкритично, но поправить стоит.

Helvetii951 commented on ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,129 @@
+            return
+        }
+        task = URLSession.shared.objectTask(for: request) { (result: Result<[PhotoResult], Error>) in
+            DispatchQueue.main.async {
+                self.isFetching = false
+                self.task = nil
+            }
```
Здесь необязательно применять `DispatchQueue.main.async`. Это ни на что не повлияет.

И желательно все обнулять в конце. Было бы классно обернуть эти строчки в [defer](https://www.avanderlee.com/swift/defer-usage-swift/). Это гарантирует их исполнение в самом конце замыкания.

Helvetii951 commented on ImagesList/ImagesListViewController.swift:
```diff
@@ -25,71 +36,145 @@ final class ImagesListViewController: UIViewController, UITableViewDelegate, UIT
+            )
+        }
+        
+        let dateFormatter: DateFormatter = {
+            let formatter = DateFormatter()
+            formatter.dateFormat = "dd MMMM yyyy"
+            formatter.locale = Locale(identifier: "ru_RU")
+            return formatter
+        }()
```
Лучше не привзваться к конкретной локали. Пусть отображение даты будет отличаться от дизайна, но пользователь увидит дату в привычном для себя виде.

Helvetii951 commented on ImagesList/ImagesListCell.swift:
```diff
@@ -1,14 +1,34 @@
-
     @IBOutlet weak var customImageView: UIImageView!
     @IBOutlet weak var likeButton: UIButton!
     @IBAction func likeButtonTapped(_ sender: Any) {
         let isLiked = likeButton.currentImage == UIImage(named: "likeButtonOn.jpeg")
         let likeImage = isLiked ? UIImage(named: "likeButtonOff.jpeg") : UIImage(named: "likeButtonOn.jpeg")
         likeButton.setImage(likeImage, for: .normal)
     }
```
Эта функция не привязана к Main.storyboard. Лучше ее убрать.

AntonDemidenk0 committed:
Исправлены замечания после ревью:
- Все форматтеры созданы один раз для использования.
- для получения даты createdAt: Date из PhotoResult\.createdAt использован ISO8601DateFormatter
- Формирование URLRequest'ов перенесено в Network Service
